### ISSUE TITLE
CA1822 - Indexers cannot be marked as static

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn-analyzers#4651]
Using .NET 5.0 in VS2019 16.8.3:

```csharp
class MyDict
{
    // CA1822	Member 'this[]' does not access instance data and can be marked as static
    public int this[int x] => 42;
}
```

More version info:
```
C# Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
```

**Comment by mavasani** [comment]
I am still trying to understand what realistic user code requires to write an indexer that does not access instance data. I feel we should Won't fix this issue until we have a valid use case.

**Comment by bartdesmet** [comment]
This was caught in two independent places:

1. A configuration component that's using indexers to read settings but is backed by a global configuration facility internally (originally the registry). It doesn't need to access instance state and because C# doesn't have static indexers, it was built as `config.Settings["bar"]` where `Settings` is an object that has a `string this[string]` indexer that's implemented without needing any instance state.
2. Test code for components that reflect on types to detect the presence of indexers and their behavior. This could be regarded as not fix-worthy, though it just deducts from the "frustration budget" for folks trying to get their codebase clean when upgrading .NET versions.

There are a couple of frustrations with this:

First, the configuration use case is definitely a valid scenario, short of a breaking change to the design of the API by taking away the indexer and using static methods instead. One can argue about the test code case but there are folks (like myself) that like to have all analyzers enabled everywhere (even though in this .NET 5.0 upgrade pass only ~5% of new warnings were actionable and did truly improve code quality, it was worth it) and false positives can be frustrating.

Second, the code fix produces invalid C#, which feels like (to put it bluntly) "the C# folks don't know their own language". This is an in-box analyzer. My stance is that code fix providers should never break the code in the first place. Maybe it gets unwieldy to track all corner cases to ensure valid code gets produced at all times, but if there's a team that can do it, it should be the extended compiler and language service team to begin with. Or code fix providers could have a "dry run" mode where the suggested fix is applied in the background and if it doesn't result in new errors, it can be safely suggested. (I know this may be way too expensive in terms of perf especially in an interactive IDE setting, but it could also be a way to test analyzers and code fix providers in general; let them silently attempt to apply their fixes in the background when otherwise idle, and keep track of cases where an attempt to apply a fix resulted in new compilation errors. All such cases have a direct correlation with customer frustration.) One picture is worth a thousand words, so doesn't the screenshot below look embarrassing with a red squiggle in the suggested "fix"?

![image](https://user-images.githubusercontent.com/7324746/103566704-517aa900-4e77-11eb-88db-41489a6468ec.png)

Third, upgrading to .NET 5.0 resulted in a few 100 new warnings or suggestions, many of which are easy to address, and some of which do reveal true bugs or questionable behavior. So that's good and keeps one motivated to "plow through". But the false positives tend to be demotivational and I can see folks just giving up and disabling a bunch of stuff. Disabled stuff rarely gets re-enabled, so one effectively lost a customer (of the analyzers and code fix providers bit that is).

Fourth, some projects are multi-targeting different frameworks (for better or for worse, it's still the reality of today). In my case, I need to target .NET Framework 4.7.2 still, and am working to add support for .NET 5.0. Local suppressions have become very cumbersome now, because the down-level builds do not have the warning and then raise `IDE0079 - Remove unnecessary suppression` (code fragment from test code below):

```csharp
        class Indexy
        {
#pragma warning disable IDE0079 // Remove unnecessary suppression (only on .NET 5.0)
#pragma warning disable CA1822 // Mark static. (https://github.com/dotnet/roslyn/issues/50197)
            public int this[int x] => x * 2;
#pragma warning restore CA1822
#pragma warning restore IDE0079
        }
```

On .NET 5.0, CA1822 comes up, so IDE0079 isn't needed. Luckily it doesn't have a "Remove "Remove unnecessary suppression" suppression" rule as well. But on .NET 4.7.2, CA1822 doesn't come up, so IDE0079 is needed. All new rules result in this pattern, so I've ultimately resorted into making IDE0079 a global suppression for test projects and I'm pretty sure that chances will be low for it to get removed from the "append-only" global suppression list eventually.

```csharp
[assembly: SuppressMessage("CodeQuality", "IDE0079:Remove unnecessary suppression", Justification = "Too much of a nuisance in test code that compiles for multiple platform targets.")]
```

I'll end on a positive note. Having more and more analyzers available in the product is a great thing; with every upgrade, I like to rerun code analysis and address things to "stay on top of the warnings pile" and smack it down. Seeing former FxCop rules be there, new rules on framework-specific semantics (e.g. `ValueTask`, LINQ operators, etc.) get added, etc. certainly helps to increase confidence in the product as a whole and it helps to improve users' code in many cases. There's just a budget to [how much false positives one can tolerate](https://en.bab.la/dictionary/dutch-english/dat-is-de-druppel-die-de-emmer-doet-overlopen) and I think it's worth trying to keep that as low as possible to avoid folks resorting to drastic measures that disable a lot of good stuff just because of frustration with a collection of small nuisances.

**Comment by mavasani** [comment]
> A configuration component that's using indexers to read settings but is backed by a global configuration facility internally (originally the registry). It doesn't need to access instance state and because C# doesn't have static indexers

Is there a reason that this needs to be an indexer as opposed to a static method with a string parameter? Note that if we feel such coding style of using indexers for accessing settings backing store instead of instance data is common, @Youssef1313's current PR would indeed make sense and we can completely avoid analyzing indexers.

**Comment by bartdesmet** [comment]
This particular concrete case is more of a compat thing; it started out that way and we can't change it anymore because we'd break folks on the public API front. Also, at the time this was designed, it felt like the most natural way to access settings which are dictionary-like.

I once suggested static indexers to the LDM (back in late 2007), but that didn't go anywhere. General indexed properties didn't exist either (and now are really just COM interop artifacts from the 4.0 days), so having a central `config` object to "dot into" a variety of things, some of which can be indexed (and could later use instance state if they wish to do so), felt like the most natural way to access these key/value stores, rather than `Get` methods that happen to be static.

In fact, I'm happy we took this route after all, because it's likely that the config store will eventually use instance state (due to cross-platform concerns where some implementations may dispatch into a registry-based approach while others may use config files or stores). Changing from `static` (using methods rather than indexers) to instance would be a breaking change as well, so we've successfully avoided that by picking this design to begin with.

Whether or not it's "common" is obviously subject for debate. Lack of static indexers tends to lead to this type of APIs (for better or for worse) when the syntactic cosmetics of `[]` brackets are desirable to hint at a dictionary-like thing. Sure, this has its limitations too (no static, no `TryGet` pattern, etc.) but these are a user's API design decisions to make, and landing on indexers is sometimes that natural thing to do.

Out of curiosity, is there a design stance on "code fix providers should never suggest a fix that produces invalid code"? That part accounts for half of the frustration, and there could be a discussion on what the analyzer looks for and considers to be "questionable behavior" versus the code fix provider suggesting an invalid fix. (As in, if it's an indexer, don't bother to suggest an invalid fix, but we believe the warning is still valid and we let the user suppress it - even though `#pragma`s are the least desirable type of fixes; doing some `_ = this;` in an accessor body doesn't look appetizing for a user to write to get around the warning.) In fact, the fix goes down the rabbit hole of suggesting invalid indexer syntax all around:

![image](https://user-images.githubusercontent.com/7324746/103572210-13828280-4e81-11eb-952a-01d6380de9db.png)

**Comment by mavasani** [comment]
> Out of curiosity, is there a design stance on "code fix providers should never suggest a fix that produces invalid code"?

Yes, absolutely. Even if we continue reporting an analyzer warning with a different message here, the code fix has to be fixed to never be offered on indexers.

**Comment by bartdesmet** [comment]
Cool; glad to hear. Then it's just a matter of deciding how "common" a use case like this is.

**Comment by mavasani** [comment]
@bartdesmet Thank you for the feedback - we have taken your suggestion and excluded indexers from CA1822 analysis, so no CA1822 diagnostic or code fix will be offered for indexers.
